In [1]:
import pandas as pd
import numpy as np

#### Note :- I have used log base e for the entire assignment.

## Part 1 :-  Tennis dataset.

In [2]:
df = pd.read_csv('play_tennis.csv')

In [3]:
df

day   outlook  temp humidity    wind play
0    D1     Sunny   Hot     High    Weak   No
1    D2     Sunny   Hot     High  Strong   No
2    D3  Overcast   Hot     High    Weak  Yes
3    D4      Rain  Mild     High    Weak  Yes
4    D5      Rain  Cool   Normal    Weak  Yes
5    D6      Rain  Cool   Normal  Strong   No
6    D7  Overcast  Cool   Normal  Strong  Yes
7    D8     Sunny  Mild     High    Weak   No
8    D9     Sunny  Cool   Normal    Weak  Yes
9   D10      Rain  Mild   Normal    Weak  Yes
10  D11     Sunny  Mild   Normal  Strong  Yes
11  D12  Overcast  Mild     High  Strong  Yes
12  D13  Overcast   Hot   Normal    Weak  Yes
13  D14      Rain  Mild     High  Strong   No

### Total entropy of the data set

In [8]:
prob = df['play'].value_counts()/df['play'].value_counts().sum()

In [25]:
t_entropy = sum( -x*np.log(x) for x in prob)

In [27]:
t_entropy

0.6517565611726531

### Entropy of Each columns

In [50]:
final_entropies = {}
for col in df.columns[1:-1]:
    
    ucol=df[col].unique()
    ucol_entropy=[]
    apply=[]
    for x in ucol:
        temp = (df[df[col]==x])
        apply.append(temp.shape[0]/df.shape[0])
        p=temp['play'].value_counts()/temp['play'].value_counts().sum()
        ucol_entropy.append(sum([ -x*np.log(x) for x in p]))
    
    final_entropies[col]=np.dot(apply,ucol_entropy)
final_entropies

{'outlook': 0.48072261929232607,
 'temp': 0.6315010221774208,
 'humidity': 0.5465122114944403,
 'wind': 0.6183974457364384}

### Now, finding the IG (Information Gain) of each columns

In [58]:
ig = {}
for i in final_entropies.keys():
    ig[i]=t_entropy-final_entropies[i]
    print(f'IG of "{i}" column : {ig[i]}')
print(f'\nMaximum IG is {max(ig.values())}')

IG of "outlook" column : 0.17103394188032706
IG of "temp" column : 0.02025553899523236
IG of "humidity" column : 0.10524434967821283
IG of "wind" column : 0.033359115436214726

Maximum IG is 0.17103394188032706


#### We can clearly see that "outlook" column has the max IG. Therefore, the first split will be done on "Outlook" column.

## Part 2 :- Iris dataset.

In [18]:
df=pd.read_csv('iris.csv')
df=df[['petal_length','species']]

In [19]:
df.rename(columns={'petal_length':'pl'},inplace=True)

In [20]:
df

pl    species
0    1.4     setosa
1    1.4     setosa
2    1.3     setosa
3    1.5     setosa
4    1.4     setosa
..   ...        ...
145  5.2  virginica
146  5.0  virginica
147  5.2  virginica
148  5.4  virginica
149  5.1  virginica

[150 rows x 2 columns]

### Total entropy of the dataset

In [21]:
prob = df['species'].value_counts()/df['species'].value_counts().sum()

In [22]:
t_entropy = sum( (-x*np.log(x)) for x in prob)

In [26]:
t_entropy

1.0986122886681096

In [24]:
## Sorting the dataset.....
df.sort_values(by='pl',ascending=True,inplace=True)

In [25]:
df

pl    species
22   1.0     setosa
13   1.1     setosa
14   1.2     setosa
35   1.2     setosa
36   1.3     setosa
..   ...        ...
131  6.4  virginica
105  6.6  virginica
117  6.7  virginica
122  6.7  virginica
118  6.9  virginica

[150 rows x 2 columns]

In [44]:
def find_IG(n):
    df1=df[df['pl']>n]
    df2=df[df['pl']<=n]
    
    if((df1[df1['species']=='versicolor']['species'].shape[0]>0) & (df1[df1['species']=='virginica']['species'].shape[0]>0) & (df1[df1['species']=='setosa']['species'].shape[0]>0) & (df2[df2['species']=='versicolor']['species'].shape[0]==0) & (df2[df2['species']=='virginica']['species'].shape[0]==0) & (df2[df2['species']=='setosa']['species'].shape[0]>0)):
        
        E1=round(-(df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='virginica'].shape[0])/(df1.shape[0]))-(df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0]))-(df1[df1['species']=='setosa'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='setosa'].shape[0])/(df1.shape[0])))
        E2=round(-(df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])))
        
    elif((df1[df1['species']=='versicolor']['species'].shape[0]>0) & (df1[df1['species']=='virginica']['species'].shape[0]>0) & (df1[df1['species']=='setosa']['species'].shape[0]==0) & (df2[df2['species']=='versicolor']['species'].shape[0]==0) & (df2[df2['species']=='virginica']['species'].shape[0]==0) & (df2[df2['species']=='setosa']['species'].shape[0]>0)):
        
        E1=round(-(df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='virginica'].shape[0])/(df1.shape[0]))-(df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0])))
        E2=round(-(df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])))
           
    elif((df1[df1['species']=='versicolor']['species'].shape[0]>0) & (df1[df1['species']=='virginica']['species'].shape[0]>0) & (df1[df1['species']=='setosa']['species'].shape[0]==0) & (df2[df2['species']=='versicolor']['species'].shape[0]>0) & (df2[df2['species']=='virginica']['species'].shape[0]==0) & (df2[df2['species']=='setosa']['species'].shape[0]>0)):
        
        E1=round(-(df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0]))-(df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])))
        E2=round(-(df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0]))-(df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])))

    elif((df1[df1['species']=='versicolor']['species'].shape[0]>0) & (df1[df1['species']=='virginica']['species'].shape[0]>0) & (df1[df1['species']=='setosa']['species'].shape[0]==0) & (df2[df2['species']=='versicolor']['species'].shape[0]>0) & (df2[df2['species']=='virginica']['species'].shape[0]>0) & (df2[df2['species']=='setosa']['species'].shape[0]>0)):
        
        E1=round(-(df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='versicolor'].shape[0])/(df1.shape[0]))-(df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])))
        E2=round(-(df2[df2['species']=='virginica'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='virginica'].shape[0])/(df2.shape[0]))-(df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0]))-(df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])))     
        
    elif((df1[df1['species']=='versicolor']['species'].shape[0]==0) & (df1[df1['species']=='virginica']['species'].shape[0]>0) & (df1[df1['species']=='setosa']['species'].shape[0]==0) & (df2[df2['species']=='versicolor']['species'].shape[0]>0) & (df2[df2['species']=='virginica']['species'].shape[0]>0) & (df2[df2['species']=='setosa']['species'].shape[0]>0)):       
        E1=round(-(df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])*np.log((df1[df1['species']=='virginica'].shape[0])/(df1.shape[0])))
        E2=round(-(df2[df2['species']=='virginica'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='virginica'].shape[0])/(df2.shape[0]))-(df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='versicolor'].shape[0])/(df2.shape[0]))-(df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])*np.log((df2[df2['species']=='setosa'].shape[0])/(df2.shape[0])))
        
    apply=(df1.shape[0]/df.shape[0])*E1+(df2.shape[0]/df.shape[0])*E2
    
    ig=round(t_entropy-apply,2)
    
    return ig

In [47]:
ig=[]
for n in df['pl']:  
    try:  
        ig.append(find_IG(n))        
    except:
        ig.append(np.nan)

In [49]:
## Adding an IG column into dataset.
df['IG']=ig

In [50]:
df

pl    species    IG
22   1.0     setosa  0.11
13   1.1     setosa  0.11
14   1.2     setosa  0.13
35   1.2     setosa  0.13
36   1.3     setosa  0.17
..   ...        ...   ...
131  6.4  virginica  0.13
105  6.6  virginica  0.12
117  6.7  virginica  0.11
122  6.7  virginica  0.11
118  6.9  virginica   NaN

[150 rows x 3 columns]

In [54]:
df[df['IG']==df['IG'].max()]

pl     species    IG
69  3.9  versicolor  0.51
82  3.9  versicolor  0.51
59  3.9  versicolor  0.51

## Therefore, the first split will be done at pl>3.9